sdk/search/Azure.Search.Documents/tests/Batching/AsyncEventExtensionsTests.cs

```diff
@@ -0,0 +1,575 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Search.Documents.Batching;
+using NUnit.Framework;
+
+namespace Azure.Search.Documents.Tests
+{
+    public class AsyncEventExtensionsTests
+    {
+        private class TestHandler
+        {
+            public TimeSpan? Delay { get; set; }
+            public string Throws { get; set; }
+
+            public EventArgs LastEventArgs { get; private set; }
+            public bool Raised => RaisedCount > 0;
+            public int RaisedCount { get; private set; } = 0;
+            public bool Completed => CompletedCount > 0;
+            public int CompletedCount { get; private set; } = 0;
+
+            public async Task Handle(EventArgs e, CancellationToken cancellationToken)
+            {
+                LastEventArgs = e;
+                RaisedCount++;
+                if (Delay != null)
+                {
+                    await Task.Delay(Delay.Value, cancellationToken);
+                }
+                if (Throws != null)
+                {
+                    throw new InvalidOperationException(Throws);
+                }
+                cancellationToken.ThrowIfCancellationRequested();
+                CompletedCount++;
+            }
+        }
+
+        private async Task Pause(Action action, TimeSpan? delay = null)
+        {
+            await Task.Delay(delay ?? TimeSpan.FromMilliseconds(10));
+            action();
+        }
+
+        [Test]
+        public void AddHandler()
+        {
+            Func<EventArgs, CancellationToken, Task> handler = null;
+            TestHandler test = new TestHandler();
+            handler += test.Handle;
+        }
+
+        [Test]
+        public void AddHandler_Null()
+        {
+            Func<EventArgs, CancellationToken, Task> handler = null;
+            handler += null;
+        }
+
+        [Test]
+        public void AddHandler_Multiple()
+        {
+            Func<EventArgs, CancellationToken, Task> handler = null;
+            handler += new TestHandler().Handle;
+            handler += new TestHandler().Handle;
+            handler += new TestHandler().Handle;
+            handler += new TestHandler().Handle;
```

Comment:
Not asserting anything here. Just making sure it doesn't throw or something?

Reply:
Yep - just making sure nothing breaks.  I started with my own `AsyncEventHandlerStore` before eventually simplifying to `Func<...>` but left the tests.